### PR TITLE
Issue #6743: fixed unused param not reported when missing description

### PR DESCRIPTION
**Pull request by Tirex**
Added one regexp to check params without docs and test from issue #6743, fixed failed tests, removed unused code.

Checkstyle tester diff report for this branch available here
https://tirex.github.io/checkstyle/diff/

### CODE REVIEW

**Comment by rnveach** [comment]
wercker failure in orekit:
> [INFO] There is 1 error reported by Checkstyle 8.26-SNAPSHOT with /pipeline/source/.ci-temp/Orekit/checkstyle.xml ruleset.
[ERROR] src/main/java/org/orekit/propagation/numerical/DSConverter.java:[54,8] (javadoc) JavadocMethod: Unused @param tag for 'state'.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------

@Tirex Please look into the issue and if it is correct, please make a PR to their repo to correct it so our CI passes.
Since their fork isn't stable, we may have to make our own fork from a stable point and add the fix.
https://github.com/checkstyle/checkstyle/blob/master/.ci/wercker.sh#L44-L47

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java [hunk 1]
```diff
@@ -121,6 +117,10 @@
     /** Compiled regexp to match Javadoc tags with no argument and {}. */
     private static final Pattern MATCH_JAVADOC_NOARG_CURLY =
             CommonUtil.createPattern("\\{\\s*@(inheritDoc)\\s*\\}");
+    /** Compiled regexp to match Javadoc tags with argument but with mussing description. */
+    private static final Pattern MATCH_JAVADOC_ARG_MISSING_DESCRIPTION =
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java]: Why was this moved outside it's original position at line 102?
If you want thing in alpabetical order, shouldn't it be there?

**Comment by Tirex** [comment]
> wercker failure in orekit:
> 
> > [INFO] There is 1 error reported by Checkstyle 8.26-SNAPSHOT with /pipeline/source/.ci-temp/Orekit/checkstyle.xml ruleset.
> > [ERROR] src/main/java/org/orekit/propagation/numerical/DSConverter.java:[54,8] (javadoc) JavadocMethod: Unused @param tag for 'state'.
> > [INFO] ------------------------------------------------------------------------
> > [INFO] BUILD FAILURE
> > [INFO] ------------------------------------------------------------------------
> 
> @Tirex Please look into the issue and if it is correct, please make a PR to their repo to correct it so our CI passes.
> Since their repo isn't stable, we may have to make our own fork from a stable point and add the fix.
> https://github.com/checkstyle/checkstyle/blob/master/.ci/wercker.sh#L44-L47

Yes, it was error in their repo.
Commit was made in dev branch
https://github.com/CS-SI/Orekit/commit/973d1bd3e4532d2c1d3a7c28136e2713bd057542.

**Comment by rnveach** [comment]
@Tirex Please update wercker then so it can be green. You can use the commit you just mentioned, but if it doesn't make CI green, you may have to use a different commit, possibly your own forked version.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java (hunk 1 above)
Comment (Tirex) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java]: Yes, you right. Rearranged code.

**Comment by Tirex** [comment]
> > https://tirex.github.io/checkstyle/diff/
> 
> Please expand regression to all projects in `projects-to-test-on.properties`
> https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#projects-to-test-onproperties

updated diff report on all projects
https://tirex.github.io/checkstyle/diff/

**Comment by rnveach** [comment]
@Tirex 
https://tirex.github.io/checkstyle/diff/apache-ant/index.html
You are not checking differences on `ThrowsCount`. The module being changed in this PR is `JavadocMethod`. You need to modify the config you are giving the regression for that check.
See https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#configuration-file

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheckTest.java [hunk 2]
```diff
@@ -115,9 +115,6 @@ public void testTags() throws Exception {
             "78:8: " + getCheckMessage(MSG_UNUSED_TAG, "@param", "Unneeded"),
             "79: " + getCheckMessage(MSG_UNUSED_TAG_GENERAL),
             "87:8: " + getCheckMessage(MSG_DUPLICATE_TAG, "@return"),
-            "109:23: " + getCheckMessage(MSG_EXPECTED_TAG, "@param", "aOne"),
-            "109:55: " + getCheckMessage(MSG_EXPECTED_TAG, "@param", "aFour"),
-            "109:66: " + getCheckMessage(MSG_EXPECTED_TAG, "@param", "aFive"),
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheckTest.java]: Check originally reported parameters with missing descriptions but now with new changes we won't even get a warning that the descriptions are missing. I personally not a fan in this lost of functionality but this is similar behavior with `@throws` tag. Since we feel this check does too much already, this type of functionality should probably be in a separate check.

@romani Are you ok with this?

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmethod/InputJavadocMethodParamsTags.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+  /**
+   * A method with an undocumented param and a missing param.
+   * Checkstyle should at least
+   * missingParam, ideally would be nice to have a flag to report
+   * also undocumentedParam
+   *
+   * @param param1 some text
+   * @param unexpectedParam
+   * @return
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmethod/InputJavadocMethodParamsTags.java]: Please add an example of a 2nd parameter where the description is on the next line and not the same line as `@param`.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheckTest.java (hunk 2 above)
Comment (Tirex) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheckTest.java]: @rnveach after implementing  #6746 this warning will be shown again.

**Comment by Tirex** [comment]
> @Tirex
> https://tirex.github.io/checkstyle/diff/apache-ant/index.html
> You are not checking differences on `ThrowsCount`. The module being changed in this PR is `JavadocMethod`. You need to modify the config you are giving the regression for that check.
> See https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#configuration-file

got some errors on the regression test. Working on it.

**Comment by romani** [comment]
Please do not forget to put all I interesting corner cases to Inputs. We should not rely on regression testing that much , we should grow our Inputs code base.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmethod/InputJavadocMethodParamsTags.java (hunk 3 above)
Comment (Tirex) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmethod/InputJavadocMethodParamsTags.java]: added more tests

**Comment by Tirex** [comment]
new report on  d4d52ed 
https://tirex.github.io/checkstyle/diff/

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java [hunk 4]
```diff
@@ -98,14 +98,14 @@
...
     private static final Pattern MATCH_JAVADOC_ARG = CommonUtil.createPattern(
             "^\\s*(?>\\*|\\/\\*\\*)?\\s*@(throws|exception|param)\\s+(\\S+)\\s+\\S*");
-
-    /** Compiled regexp to match first part of multilineJavadoc tags. */
-    private static final Pattern MATCH_JAVADOC_ARG_MULTILINE_START = CommonUtil.createPattern(
-            "^\\s*(?>\\*|\\/\\*\\*)?\\s*@(throws|exception|param)\\s+(\\S+)\\s*$");
+    /** Compiled regexp to match Javadoc tags with argument but with missing description. */
+    private static final Pattern MATCH_JAVADOC_ARG_MISSING_DESCRIPTION =
+        CommonUtil.createPattern("^\\s*(?>\\*|\\/\\*\\*)?\\s*@(throws|exception|param)\\s+"
+            + "(\\S[^*]*)(?:(\\s+|\\*\\/))?");
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java]: `(\\S+)\\s*$` is converted to `(\\S[^*]*)(?:(\\s+|\\*\\/))?`.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheckTest.java (hunk 2 above)
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheckTest.java]: After thinking for a while ...
I think it is ok to have such regression, it is good that it is false-negative, that is good.
It is good to validate params that are referenced in javadoc, at least user tried to explain it, so it is good help him to it properly.
in #6746, it should be configurable to report on missed parameters, as most of parameters are self-explaining, so it is ok to skip for some time.

**Comment by romani** [comment]
@Tirex , please rebase on latest master code, to make sure that validation on Orekit is passed.
some update for Orekit was just merged tocheckstyle master.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java (hunk 4 above)
Comment (Tirex) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java]: (\\S+)\\s*$ expression did not cover this expression case:
 /** @param x*/
so left unchanged
